server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -1843,6 +1891,7 @@ public static class Builder {
         private Integer routingNumShards;
         private boolean isSystem;
         private IndexLongFieldRange timestampRange = IndexLongFieldRange.NO_SHARDS;
+        private IndexLongFieldRange eventIngestedRange = IndexLongFieldRange.NO_SHARDS;
```

Comment:
I tried doing that early on in this PR when bwc tests started failing: https://github.com/elastic/elasticsearch/pull/106252#discussion_r1592835564

However, @DaveCTurner argued instead that we need to default to NO_SHARDS and assign (and assert for) UNKNOWN in the mixed version scenarios.

> it is relatively hard to reason about having caught all places with the current default

Very true. That's why it took me a long time to track down all the places that bwc and mixed cluster tests were failing (it was/is quite opaque to me, since I don't understand all of the close/restore/open/create/recover etc pathways that converge on this). I'm still not convinced I haven't missed any edge cases.

Because of that I'm relying on es-distributed to determine what's the right solution here and what the edge cases area.

Reply:
From that comment:
> I think we should also have an assertion somewhere that all these values are UNKNOWN until all nodes in the cluster support the new feature.

Did we add such an assertion? I suppose we can add that when building cluster state maybe? That would be just as good I think.

I think getting to a NO_SHARDS default is right, once we remove BWC logic. If adding the above assertion is infeasible, I'd probably prefer to make it null default here and assert that it has an explicit value set (which could be from the prior index metadata definition).

Reply:
> Did we add such an assertion? I suppose we can add that when building cluster state maybe? That would be just as good I think.

Asserts for UNKNOWN are done in two places:

https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1693

https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R1855

Both of which are in `writeTo` methods. It is not clear to me where else to add them.

And with the [first new commit](https://github.com/elastic/elasticsearch/commit/0e69ac3c7785499cbf3de6977d2c5f414aaebe9b) we now pass in minClusterTransportVersion to IndexMetadata.withTimestampRanges to ensure UNKNOWN is set when there is still a mixed-cluster.

> I suppose we can add that when building cluster state maybe? That would be just as good I think.

[The second new commit](https://github.com/elastic/elasticsearch/pull/106252/commits/11dca2acb2b313735dad45a0b322ebcabdc78842) sort of addresses that. The problem is that `IndexMetadata.build` and `IndexMetadataDiff.build` are called from a lot of places, some of which don't have access to the min cluster transport version. Example: `fromXContent` methods. That would be needed in order to do asserts.

So instead I tried to require passing in `TransportVersion minClusterTransportVersion` into the `Builder#eventIngestedRange`. In many cases, the calling function has access to the minClusterTransportVersion but in a few places (again like fromXContent), they do not, so I pass in `null` in those cases, making this only a partial solution.

Let me know if you think if that is good enough or we need something more strict. Or maybe I'm missing a better way to do it.

> I think getting to a NO_SHARDS default is right, once we remove BWC logic. If adding the above assertion is infeasible, I'd probably prefer to make it null default here and assert that it has an explicit value set (which could be from the prior index metadata definition).

I tried setting the default to the `null`. That causes over 5,200 unit tests to fail due to [this assert](https://github.com/elastic/elasticsearch/pull/106252/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R2340) which I added a while back. So I don't think's feasible?

Reply:
About cluster state building, I imagined just doing this in `ClusterState.Builder.build` or in the ClusterState constructor. It can simply loop over all the indices checking that the event ingested range is UNKNOWN. No validation is needed once the min transport version is above a certain version number. I think this is easily done, so I'd like to have it even if we strengthened it elsewhere.

Reply:
Oh, good idea. Added in next push.

Reply:
Hmm, I still hit a problem with this approach.

First I tried adding the assert (see below) to the second ClusterState constructor. That causes >900 unit tests to fail with errors like:

```
java.lang.AssertionError: event.ingested range should be UNKNOWN but is NO_SHARDS for index: [idx] minTransportVersion: 7170099
```

I then tried adding to the `ClusterState` builder. Same problem (and I have to compute the min version rather than get it for free).

I'm not sure if all these tests failing is indicative of a major problem with the production code or just an artifact of how unit tests are constructed. I ran three IT tests that I know interact to some degree with my code and they had mixed results

* `FrozenIndexIT` passed
* `SearchableSnapshotsCanMatchOnCoordinatorIntegTests` - failed (same error message as shown above)
* `SearchableSnapshotsIntegTests` - some passed, some failed

Does this make you think we should set the default to UNKNOWN instead?

Here's the code I added (this is the variant for adding it to the builder):

```java
            assert assertEventIngestedIsUnknownInMixedClusters(metadata,
                CompatibilityVersions.minimumVersions(compatibilityVersions.values()));

...

        private boolean assertEventIngestedIsUnknownInMixedClusters(Metadata metadata, CompatibilityVersions compatibilityVersions) {
            if (compatibilityVersions.transportVersion().before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
                for (IndexMetadata indexMetadata : metadata.indices().values()) {
                    assert indexMetadata.getEventIngestedRange() == IndexLongFieldRange.UNKNOWN
                        : "event.ingested range should be UNKNOWN but is "
                        + indexMetadata.getEventIngestedRange() + " for index: " + indexMetadata.getIndex()
                        + " minTransportVersion: " + previous.getMinTransportVersion();
                }
            }
            return true;
        }

```

Reply:
I think we should solve this by relaxing it to allow both `NO_SHARDS` and `UNKNOWN`. We should then do the same in the `writeTo` assertion.

Reply:
Hmm, I take that back (for now at least). We should be able to make this work I think. The main problem seems to be the compatiblity version which is not always initialized. I think following diff solves at least one such case:

```
diff --git a/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java b/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
index 8d72715219e..0d057db3c03 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java
@@ -172,9 +172,9 @@ public class TransportClusterStateAction extends TransportMasterNodeReadAction<C
 
         if (request.nodes()) {
             builder.nodes(currentState.nodes());
-            builder.nodeIdsToCompatibilityVersions(getCompatibilityVersions(currentState));
             builder.nodeFeatures(getClusterFeatures(currentState));
         }
+        builder.nodeIdsToCompatibilityVersions(getCompatibilityVersions(currentState));
         if (request.routingTable()) {
             if (request.indices().length > 0) {
                 RoutingTable.Builder routingTableBuilder = RoutingTable.builder();
```

I'll experiement a bit more with some unittests too.

Reply:
Doing it in the constructor like this:

```
        assert compatibilityVersions.isEmpty() || assertEventIngestedIsUnknownInMixedClusters(metadata, this.minVersions);
    }

    private boolean assertEventIngestedIsUnknownInMixedClusters(Metadata metadata, CompatibilityVersions compatibilityVersions) {
        if (compatibilityVersions.transportVersion().before(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
            for (IndexMetadata indexMetadata : metadata.indices().values()) {
                assert indexMetadata.getEventIngestedRange() == IndexLongFieldRange.UNKNOWN
                    : "event.ingested range should be UNKNOWN but is "
                    + indexMetadata.getEventIngestedRange() + " for index: " + indexMetadata.getIndex()
                    + " minTransportVersion: " + compatibilityVersions.transportVersion();
            }
        }
        return true;
    }
```

should also work I think.

Reply:
Had to change it to:

```
        assert compatibilityVersions.isEmpty()
            || blocks.hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)
            || assertEventIngestedIsUnknownInMixedClusters(metadata, this.minVersions);
```

and change `eventTimestampRange` to `UNKNOWN` in `ClusterRerouteResponseTests.createClusterState` and `ClusterStateTests.buildClusterState` (and fixing it in the 3 toXContent tests)

Reply:
Yes, sorry. I missed this set of comments in your previous review. I've confirmed that your suggested assert works against server unit tests (with the slight test adjustments you mentioned) and the handful of IT tests I tried. Will be in next push to see if it passes CI and bwc tests.